articles/cross_compilation_build_tools_design.md

```diff
@@ -0,0 +1,161 @@
+However, that tutorial doesn't provide instructions to build other ROS 2 packages, but just for building a whole ROS 2
+distribution. Ideally, we should be able to cross-compile a ROS package for ROS 2 by launching a single `colcon build`
+command. Finally, cross-compiling for other platforms besides ARM64 is not supported in the
+ [ros2/cross_compile repository](https://github.com/ros2/cross_compile/).
+
+## Proposed Approach
+
+We propose continuing with the general approach of the cross-compilation tutorial, based on building a sysroot for
+the target platform using qemu and Docker, and then using `cmake-toolchains` with a toolchain file pointing `CMAKE_SYSROOT` to the sysroot
+location to compile with C and C++ cross-compilers.
+
+On top of that we propose using a **colcon plugin `cc-build`** to wrap the instructions on the cross-compilation tutorial into a single
+build command. That plugin would define a new `colcon_core.verb` entry point pointing to a class `CCBuildVerb` that uses the
+[Decorator pattern](https://en.wikipedia.org/wiki/Decorator_pattern) to wrap the [`BuildVerb` class](https://github.com/colcon/colcon-core/blob/master/colcon_core/verb/build.py#L74)
+used by `colcon build` with additional functionality to handle the sysroot and toolchain file, and to inject additional `cmake-args` arguments.
+That would allows us to run the single command below from a ROS 2 overlay workspace on a developer workstation, for cross-compiling
+the packages in the workspace up to a package `performance_test` for the ARM-HF platform.
+
+```bash
+colcon cc-build --arch armhf --os ubuntu_bionic \
+  --packages-up-to performance_test # other `colcon build` arguments
+```
+
+That would should result on a new workspace directory `armhf-ubuntu_bionic` with subdirectories `build`, `install`, and `log` directories, with the 
+ross-compiled binaries. All options for `colcon build` will be supported by forwarding them to the `BuildVerb` class. That would be the only command required for cross-compiling the packages in a workspace. Under the hood, the plugin would do the following:
+
+- The [development tools described in the cross-compilation tutorial](https://index.ros.org/doc/ros2/Tutorials/Cross-compilation/#install-development-tools)
+ (`qemu-user-static`, `g++-aarch64-linux-gnu`, ..) should be installed in the development workstation used for the build. The `cc-build` plugin will
+ declare those system dependencies using a `stdeb.cfg` file like is done on [other colcon plugins](https://github.com/colcon/colcon-core/blob/master/stdeb.cfg).
+- In order to build the sysroot, the plugin downloads a base ROS 2 Docker image for the target platform, and builds a workspace dependent sysroot image by doing a `COPY` to copy the contents of the workspace into the container, and running `rosdep` on the copy. That ensures no workspace system dependency is missing. Note we we can run `rosdep`
```

Comment:
Thanks for letting me know about the typo, it's fixed in a later commit.

Regarding the list of supported platforms, after some feedback in this PR the idea is that instead of a colcon plugin we'd add a command for setting up the sysroot, and a colcon mixin for adding the required cross compilation flags to `colcon build`. For example we might build ROS 2 for armhf as follows

```bash
# Setup the sysroot for the platform (generic_armhf, ubuntu_bionic, ROS crystal, RMW FastRTPS)
# this leaves the sysroot at generic_armhf-ubuntu_bionic-fastrtps-crystal/sysroot
cc-setup-sysroot --arch generic_armhf --os ubuntu_bionic --distro crystal --rmw fastrtps \
   --sysroot-base-image ros2:armhf_bionic_crystal_base-crystal
# this creates some symlinks like /lib/arm-linux-gnueabihf -> $(pwd)/generic_armhf-ubuntu_bionic-fastrtps-crystal/sysroot/lib/arm-linux-gnueabihf in line with the "Hack to find Poco" from https://github.com/ros2/cross_compile/blob/master/entry_point.sh#L38
bash generic_armhf-ubuntu_bionic-fastrtps-crystal/cc_system_setup.bash

# this setups some env vars for cross compilation
source generic_armhf-ubuntu_bionic-fastrtps-crystal/cc_build_setup.bash

# Run colcon build using the mixin for this platform
colcon build --mixin cc-generic_armhf-ubuntu_bionic-fastrtps-crystal
```

The image `ros2:armhf_bionic_crystal_base-crystal` above refers to a Docker image with just the basics depedencies to build ROS2. To cross compile a workspace with additional packages that are not part of ROS 2, we use a different argument for `--sysroot-base-image` that corresponds to an image with ROS 2 pre built for that architecture. Then `cc_build_setup.bash` would source `generic_armhf-ubuntu_bionic-fastrtps-crystal/sysroot/opt/ros/crystal/setup.bash` as part of the setup.

This means we'd have a different colcon mixin for each platform.

Reply:
> This means we'd have a different colcon mixin for each platform.

If the duplication between them becomes an issue we could consider supporting one mixin to extend another mixin - just a thought.